### PR TITLE
Calculate the text alignment offset

### LINKED ISSUE #5834: Bounding text width adds extra top padding
## Bevy version

0.8

## Relevant system information

Win10 / Vulkan

```ignore
AdapterInfo { name: "NVIDIA GeForce RTX 2070", vendor: 4318, device: 7938, device_type: DiscreteGpu, backend: Vulkan }
```

## What you did

To work around #1490, constrain the size of a text in an attempt to wrap it:

```rust
                .spawn_bundle(NodeBundle {
                    style: Style {
                        flex_direction: FlexDirection::ColumnReverse, // top to bottom
                        align_items: AlignItems::FlexStart,           // align content top
                        ..default()
                    },
                    color: UiColor(Color::rgba_u8(122, 47, 125, 214)),
                    ..default()
                })
                .with_children(|parent| {
                    parent
                        .spawn_bundle(TextBundle {
                            text: Text::from_section(
                                "",
                                TextStyle {
                                    font: font.clone(),
                                    font_size: 17.0,
                                    color: Color::GRAY,
                                },
                            ),
                            style: Style {
                                size: Size::new(Val::Px(490.), Val::Auto), // #1490 - Text wrapping should work with a percent width
                                ..default()
                            },
                            ..default()
                        });
                });
```

## What went wrong

Without the size, or with a percentage (#1490):

image

With the code above:

image

But now with a shorter text and the same size constraint, some top padding appeared:

image

And it's proportional to the constraint width, here's for `370.` instead of `490.`:

image

When the constraint reaches the actual calculated text size, the padding disappear. When going further below, the text as expected starts to wrap again.

## Additional information

The text vertical alignment is ignored, so cannot even work around by vertically aligning to top. I'm not sure with the size is counted toward the text wrapping, since the size is that of the node, whereas text wrapping should work with the actual calculated size of the glyphs.

### LINKED ISSUE #6717: TextAlignment doesn't work as intended
## Bevy version

0.9

## What you did

Attempted to center word wrapped text.

## What went wrong

The text was centered in a space as wide as the longest line, rather than the width of the text node, and then moved to the left side of the text node. This resulted in it being off-center to the left.

## Additional information

It appears that TextAlignment first calculates the dimensions of the text itself, ignoring the any size specified on the text node. Then it aligns the text within that box, and puts that box against the top-left of the box of the text node. This means that when there is no word wrap, TextAlignment will do nothing at all, because it will be aligning the text in a box that's exactly the same size as the text.

When there is word wrap, it'll determine the width of the box by the width of the widest line. Then all the shorter lines will be aligned correctly relative to that. However, the longest line will generally still be less wide than the width of the text node, so the extra space will be made up with padding on the right.

The vertical alignment portion effectively never works, presumably for the same reason. The vertical size of the text is calculated, and it's vertically aligned in that space, which doesn't do anything. And then if the text node is taller than that, the extra space is made up with bottom padding.

In this screenshot, the text node's dimensions are the dimensions of the gray rectangle, and it's given a `TextAlignment::CENTER`.
image

The relevant bit of my UI code that produced that screenshot is this:
```
parent
	.spawn(NodeBundle {
		style: Style {
			size: Size::new(Val::Percent(100.0), Val::Px(180.0)),
			padding: UiRect::all(Val::Px(4.0)),
			margin: UiRect { top: Val::Px(10.0), ..default() },
			flex_direction: FlexDirection::Column,
			..default()
		},
		background_color: BackgroundColor(Color::rgb(0.3, 0.3, 0.3)),
		..default()
	})
	.with_children(|parent| {
		parent
			.spawn(
				TextBundle::from_section(
					&scroll.description,
					TextStyle {
						font: asset_server.load("fonts/FiraSans-Bold.ttf"),
						font_size: 20.0,
						color: Color::WHITE,
					}
				)
				.with_style(Style {
					size: Size::new(Val::Px(272.0), Val::Px(172.0)),
					..default()
				})
				.with_text_alignment(TextAlignment::CENTER)
			);
	});
```

### PR DESCRIPTION
**Pull request by ickshonpe**
# Objective

Fixes  #5834, #6717

Related to #5502

The alignment offset of text is never calculated. Text is translated minus half the node size regardless of what the `TextAlignment` is set to.

## Solution

Calculate the text alignment offset.

## Changelog
Added text alignment offset calculations

### CODE REVIEW

**Review by alice-i-cecile — APPROVED** [review]
Very nice style, and a vital bug fix.

File: crates/bevy_ui/src/render/mod.rs [hunk 1]
```diff
@@ -326,9 +326,19 @@ pub fn extract_text_uinodes(
...
+            let alignment_offset = (match text.alignment.vertical {
+                VerticalAlign::Center => -0.5 * text_layout_info.size.y,
+                VerticalAlign::Top => -0.5 * uinode.size().y,
+                VerticalAlign::Bottom => 0.5 * uinode.size().y - text_layout_info.size.y,
+            } * Vec3::Y)
+                + (match text.alignment.horizontal {
+                    HorizontalAlign::Left => -0.5 * uinode.size().x,
+                    HorizontalAlign::Center => -0.5 * text_layout_info.size.x,
+                    HorizontalAlign::Right => 0.5 * uinode.size().x - text_layout_info.size.x,
+                } * Vec3::X);
```
Comment (mockersf) [inline crates/bevy_ui/src/render/mod.rs:339]:
```suggestion
            let alignment_offset = Vec2::new(
                match text.alignment.horizontal {
                    HorizontalAlign::Left => -0.5 * uinode.size().x,
                    HorizontalAlign::Center => -0.5 * text_layout_info.size.x,
                    HorizontalAlign::Right => 0.5 * uinode.size().x - text_layout_info.size.x,
                },
                match text.alignment.vertical {
                    VerticalAlign::Center => -0.5 * text_layout_info.size.y,
                    VerticalAlign::Top => -0.5 * uinode.size().y,
                    VerticalAlign::Bottom => 0.5 * uinode.size().y - text_layout_info.size.y,
                },
            )
            .extend(0.0);
```

Comment (ickshonpe) [reply crates/bevy_ui/src/render/mod.rs:339]: Oh yes, definitely cleaner.

**Comment by mockersf** [comment]
could you check the difference on the example `text_debug`?

**Review by mockersf — CHANGES_REQUESTED** [review]
could you check the difference on the example `text_debug`?

**Comment by ickshonpe** [comment]
> could you check the difference on the example `text_debug`?

Yes, the output seems identical to main. All of the UI examples run correctly, as best as I can tell.

**Comment by mockersf** [comment]
It's not, the pink text is not placed the same

on main:
Screenshot 2022-11-22 at 09 23 00

on this PR:
Screenshot 2022-11-22 at 09 23 23

**Comment by ickshonpe** [comment]
Got it, I think. It is a second bug in the extract_text_uinodes function that was hidden by the first:

```rust
let extracted_transform = global_transform.compute_matrix()
                    * Mat4::from_scale(Vec3::splat(scale_factor.recip()))
                    * Mat4::from_translation(
                        alignment_offset * scale_factor + text_glyph.position.extend(0.),
                    );
```

The alignment offset is multiplied by the scale factor twice.

**Comment by ickshonpe** [comment]
> This also affects Text2d, I'll put the fix for it into a separate PR.

Text2d actually already works correctly, as it divides the text size by the scale factor first once in the layout function and then multiplies the alignment offset twice in extract_text2d_sprite.

File: crates/bevy_ui/src/render/mod.rs [hunk 2]
```diff
@@ -348,11 +361,10 @@ pub fn extract_text_uinodes(
                 let atlas_size = Some(atlas.size);
 
                 // NOTE: Should match `bevy_text::text2d::extract_text2d_sprite`
```
Comment (rparrett) [inline crates/bevy_ui/src/render/mod.rs:363]: Is this comment (and the corresponding one in `text2d.rs`) still accurate? The two chunks of code seems to have diverged quite a bit.

Comment (ickshonpe) [reply crates/bevy_ui/src/render/mod.rs:363]: It's not really good either way. If there is duplication it should probably be abstracted away.

I think delete it, but leave the comment in Text2d.rs as outside the scope of this PR.

File: crates/bevy_ui/src/render/mod.rs [hunk 3]
```diff
@@ -347,12 +360,9 @@ pub fn extract_text_uinodes(
                 let rect = atlas.textures[index];
                 let atlas_size = Some(atlas.size);
 
-                // NOTE: Should match `bevy_text::text2d::extract_text2d_sprite`
                 let extracted_transform = global_transform.compute_matrix()
                     * Mat4::from_scale(Vec3::splat(scale_factor.recip()))
-                    * Mat4::from_translation(
-                        alignment_offset * scale_factor + text_glyph.position.extend(0.),
-                    );
+                    * Mat4::from_translation(alignment_offset + text_glyph.position.extend(0.));
```
Comment (Weibye) [inline crates/bevy_ui/src/render/mod.rs:365]: is it correct that the scale_factor should no longer be considered here?

File: crates/bevy_ui/src/render/mod.rs (hunk 2 above)
Comment (ickshonpe) [reply crates/bevy_ui/src/render/mod.rs]: Looked at text2d.rs some more and fixed some issues (including removing the other comment) in #6741

File: crates/bevy_ui/src/render/mod.rs (hunk 3 above)
Comment (ickshonpe) [reply crates/bevy_ui/src/render/mod.rs:365]: I *think* so. Scale factor is applied all over the place and it's very confusing though. I'll have another look

Comment (ickshonpe) [reply crates/bevy_ui/src/render/mod.rs:365]: I added an example `text_alignment` that cycles through different TextAlignments and scale factors and it definitely looks correct. I'm worried there are other edge cases but I think that even if there are some problems it's a definite improvement on before.

File: examples/ui/text_alignment.rs [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+fn main() {
+    App::new()
+        .add_plugins(DefaultPlugins)
+        .add_startup_system(setup)
+        .add_system(update)
+        .run();
+}
+fn setup(mut commands: Commands, asset_server: Res<AssetServer>) {
+    commands.insert_resource(UiScale { scale: 0.5 });
```
Comment (Weibye) [inline examples/ui/text_alignment.rs:25]: Why is UiScale 0.5 necessary here? Does it add to the example?

File: examples/ui/text_alignment.rs [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        *t = 0.5;
+        *i = (*i + 1) % ALIGNMENTS.len();
+        let mut text = text_query.single_mut();
+        text.alignment = ALIGNMENTS[*i];
+        text.sections[0].value = format!(
+            "{:?}-{:?}",
+            text.alignment.vertical, text.alignment.horizontal
+        );
+
+        if *i % 2 == 0 {
```
Comment (Weibye) [inline examples/ui/text_alignment.rs:81]: What is this doing exactly? On every other alignment change you increase the UiScale?

**Review by Weibye — COMMENTED** [review]
The changes in bevy_ui looks good imo, but some questions on the example

File: examples/ui/text_alignment.rs (hunk 5 above)
Comment (ickshonpe) [reply examples/ui/text_alignment.rs:81]: Yes, the example is just something I hacked together, I didn't put any thought or time into it.
Its only purpose is to make sure the text displays correctly for each of the text alignment variants at scale factors below 1, at 1, and above 1. I'll remove it from the PR.

Comment (alice-i-cecile) [reply examples/ui/text_alignment.rs:81]: I actually like it in this pr, just add a comment.

File: crates/bevy_ui/src/render/mod.rs (hunk 3 above)
Comment (ickshonpe) [reply crates/bevy_ui/src/render/mod.rs]: I get it now, the whole alignment_offset doesn't need to be multiplied by the scale factor,  but the node_size does. I was missing that UiScale and scale_factor are different things. Everything should work now.

**Review by Weibye — APPROVED** [review]

**Comment by ickshonpe** [comment]
This is the wrong approach I think after all. It works but it's a hack hiding the real problems:

1. Text should only have a horizontal alignment not vertical, and that shouldn't affect the position of the text. You use Flexbox to position the text. That's how it works in the browser.
2. The measure func incorrectly sizes text nodes to fill the entire parent node when it should only take up exactly as much space as needed to fit the text.

**Pull request closed without merging**